File: SAR-Simulation/SAR.js
const cv = require('opencv4nodejs');
const { createCanvas, loadImage } = require('canvas');
const fs = require('fs');
const path = require('path');
const { fft2, ifft2, fftshift, ifftshift } = require('fft-js');

function regionGrowing(img, seed, threshold) {
    const visited = new Array(img.rows).fill(0).map(() => new Array(img.cols).fill(false));
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    const queue = [[seed.x, seed.y]];
    const seedValue = img.at(seed.y, seed.x);

    while (queue.length > 0) {
        const [x, y] = queue.shift();
        if (visited[y][x]) continue;
        if (Math.abs(img.at(y, x) - seedValue) <= threshold) {
            visited[y][x] = true;
            img.set(y, x, 0);
            for (let i = 0; i < 4; i++) {
                const nx = x + dx[i];
                const ny = y + dy[i];
                if (nx >= 0 && nx < img.cols && ny >= 0 && ny < img.rows && !visited[ny][nx]) {
                    queue.push([nx, ny]);
                }
            }
        }
    }
    return img;
}

function lowPassFilter(image, cutoff) {
    const [rows, cols] = image.sizes;
    const crow = Math.floor(rows / 2);
    const ccol = Math.floor(cols / 2);
    const mask = new cv.Mat(rows, cols, cv.CV_8U, 0);
    mask.getRegion(new cv.Rect(ccol - cutoff, crow - cutoff, cutoff * 2, cutoff * 2)).setTo(255);

    const fshift = fftshift(fft2(image));
    const fshiftMasked = fshift.map((row, i) => row.map((val, j) => val * mask.at(i, j)));
    const imgBack = ifft2(ifftshift(fshiftMasked));

    return imgBack;
}

function sar(filePath, x, y, threshold, b) {
    const img = cv.imread(filePath, cv.IMREAD_GRAYSCALE);

    const seed = { x: parseInt(x), y: parseInt(y) };
    regionGrowing(img, seed, parseInt(threshold));

    const omega = 5.0 * Math.sqrt(img.rows * img.cols) / b;

    const theta = cv.Mat.zeros(img.rows, img.cols, cv.CV_64F).map(() => Math.random() * 2 * Math.PI);
    const complexImg = cv.Mat.zeros(img.rows, img.cols, cv.CV_64F);
    img.convertTo(complexImg, cv.CV_64F);

    const complexImgWithTheta = complexImg.map((val, i, j) => val * Math.exp(1 * j * theta.at(i, j)));
    const filteredImg = lowPassFilter(complexImgWithTheta, parseInt(omega));

    const outputPath = path.join(__dirname, 'results', 'filtered_image.png');
    cv.imwrite(outputPath, cv.Mat.abs(filteredImg));

    return outputPath;
}

module.exports = sar;

File: week6/backend/SAR.py
import numpy as np
from PIL import Image
from scipy import signal
import io

def region_growing(img, seed, threshold):
    mask = np.zeros(img.shape, dtype=bool)
    seed_value = int(img[seed])
    mask[seed] = True
    neighbors = [(0,1), (1,0), (0,-1), (-1,0)]
    stack = [seed]
    while stack:
        x, y = stack.pop()
        for dx, dy in neighbors:
            nx, ny = x + dx, y + dy
            if 0 <= nx < img.shape[0] and 0 <= ny < img.shape[1]:
                if not mask[nx, ny] and abs(int(img[nx, ny]) - seed_value) <= threshold:
                    mask[nx, ny] = True
                    stack.append((nx, ny))
    return mask

def sar(image_data, x, y, threshold, b):
    # 将图像数据转换为numpy数组
    img = Image.open(io.BytesIO(image_data)).convert('L')
    img_array = np.array(img)

    # 应用区域增长算法
    mask = region_growing(img_array, (y, x), threshold)

    # 将选中区域设为黑色，其他区域保持不变
    img_array[mask] = 0

    # 计算带宽omega
    omega = 5.0 * np.sqrt(img_array.shape[0] * img_array.shape[1]) / b

    # 生成随机相位
    theta = np.random.uniform(0, 2*np.pi, img_array.shape)
    complex_img = img_array * np.exp(1j * theta)

    # 应用低通滤波器
    rows, cols = img_array.shape
    crow, ccol = rows // 2, cols // 2
    mask = np.zeros((rows, cols), dtype=bool)
    y, x = np.ogrid[-crow:rows-crow, -ccol:cols-ccol]
    mask = x*x + y*y <= omega*omega
    f = np.fft.fft2(complex_img)
    f_shifted = np.fft.fftshift(f)
    f_filtered = f_shifted * mask
    f_filtered_shifted = np.fft.ifftshift(f_filtered)
    filtered_img = np.fft.ifft2(f_filtered_shifted)

    # 计算幅度
    result = np.abs(filtered_img)

    # 归一化结果到0-255范围
    result = ((result - result.min()) / (result.max() - result.min()) * 255).astype(np.uint8)

    # 创建PIL图像
    result_img = Image.fromarray(result)

    # 将图像转换为字节流
    img_byte_arr = io.BytesIO()
    result_img.save(img_byte_arr, format='PNG')
    img_byte_arr = img_byte_arr.getvalue()

    return result